Better reporting of integration test container issues

### Description

I recently had a situation where, during integration tests with container build, my container was immediately crashing on start (mismatched jdk versions). The issue though is that there was no feedback on the Quarkus / test output side that this had occurred, just a very cryptic

```
java.lang.RuntimeException at QuarkusIntegrationTestExtension.java:387
        Caused by: java.lang.IllegalStateException at LauncherUtil.java:109
```
for the failed test. While a simple fix, was a stumper until I managed to catch the container starting then failing. Given the app I found this in is super simple and can run with out other service (database) dependencies (meaning I could just run the thing and know it should run on its own), this would be a rather annoying issue to diagnose with more complex apps.

It would be great to see a more concrete message, similar to running regular tests if your server failed to start for whatever reason. Also including container logs in the output would be helpful.

### Implementation ideas

I'm less privvy to the mechanations of how the test instance of the app container is run, but assuming it's testcontainer based, it should be possible to adjust things to be more wary of the container state. Given the bar would be "better log output", should be doable.

--- a/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
+++ b/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
@@ -19,7 +19,10 @@ import static java.util.Optional.empty;
 
 import java.io.Closeable;
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
@@ -78,6 +81,8 @@ import io.smallrye.config.SmallRyeConfigBuilder;
 public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithContextExtension
         implements BeforeTestExecutionCallback, AfterTestExecutionCallback, BeforeEachCallback, AfterEachCallback,
         BeforeAllCallback, AfterAllCallback, TestInstancePostProcessor, ParameterResolver {
+
+    private static final int APP_LOG_TAIL_LINES = 50;
 
     private static boolean failedBoot;
 
@@ -196,6 +201,7 @@ public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithCont
                     if (appLogFile.exists() && (appLogFile.length() > 0)) {
                         System.err.println("Failed to launch the application. The application logs can be found at: "
                                 + appLogFile.getAbsolutePath());
+                        printApplicationLogTail(appLogFile);
                     }
                 } catch (IllegalStateException ignored) {
 
@@ -206,6 +212,33 @@ public class QuarkusIntegrationTestExtension extends AbstractQuarkusTestWithCont
             }
         }
         return state;
+    }
+
+    // When the application fails to boot, the launcher only reports a generic "Unable to determine the
+    // status of the running process" error; the real cause is in the application log file. Echo the tail of
+    // that file to System.err so the cause is visible directly in the test output (e.g. in CI), not only on disk.
+    private static void printApplicationLogTail(File appLogFile) {
+        try {
+            applicationLogTail(Files.readAllLines(appLogFile.toPath()), APP_LOG_TAIL_LINES, appLogFile.getName())
+                    .forEach(System.err::println);
+        } catch (IOException ignored) {
+            // best-effort; the path to the full log was already printed above
+        }
+    }
+
+    // Formats the last `maxLines` lines of the given log content with a header, or returns an empty list
+    // when there is nothing to show. Kept package-private and pure so it can be unit-tested directly.
+    static List<String> applicationLogTail(List<String> lines, int maxLines, String fileName) {
+        if (lines.isEmpty()) {
+            return List.of();
+        }
+        int from = Math.max(0, lines.size() - maxLines);
+        List<String> tail = new ArrayList<>(lines.size() - from + 1);
+        tail.add("Last " + (lines.size() - from) + " line(s) of " + fileName + ":");
+        for (String line : lines.subList(from, lines.size())) {
+            tail.add("    " + line);
+        }
+        return tail;
     }
 
     private QuarkusTestExtensionState doProcessStart(Class<? extends QuarkusTestProfile> profile, ExtensionContext context)

--- a/test-framework/junit/src/test/java/io/quarkus/test/junit/QuarkusIntegrationTestExtensionTest.java
+++ b/test-framework/junit/src/test/java/io/quarkus/test/junit/QuarkusIntegrationTestExtensionTest.java
@@ -1,0 +1,41 @@
+package io.quarkus.test.junit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.stream.IntStream;
+
+import org.junit.jupiter.api.Test;
+
+public class QuarkusIntegrationTestExtensionTest {
+
+    @Test
+    public void tailReturnsLastLinesWithHeaderWhenLogExceedsLimit() {
+        List<String> lines = IntStream.rangeClosed(1, 120).mapToObj(i -> "line " + i).toList();
+
+        List<String> tail = QuarkusIntegrationTestExtension.applicationLogTail(lines, 50, "quarkus.log");
+
+        assertThat(tail).hasSize(51); // 1 header + 50 lines
+        assertThat(tail.get(0)).isEqualTo("Last 50 line(s) of quarkus.log:");
+        assertThat(tail.get(1)).isEqualTo("    line 71");
+        assertThat(tail.get(50)).isEqualTo("    line 120");
+    }
+
+    @Test
+    public void tailReturnsAllLinesWhenLogIsSmallerThanLimit() {
+        List<String> lines = List.of("first", "second", "third");
+
+        List<String> tail = QuarkusIntegrationTestExtension.applicationLogTail(lines, 50, "quarkus.log");
+
+        assertThat(tail).containsExactly(
+                "Last 3 line(s) of quarkus.log:",
+                "    first",
+                "    second",
+                "    third");
+    }
+
+    @Test
+    public void tailReturnsEmptyForEmptyLog() {
+        assertThat(QuarkusIntegrationTestExtension.applicationLogTail(List.of(), 50, "quarkus.log")).isEmpty();
+    }
+}

